pkg/node/table.go

```diff
@@ -0,0 +1,375 @@
+	GetReconciliationStatus(reconciler string) reconciler.Status
+
+	// SetReconciliationStatus sets the reconciliation status for the given
+	// reconciler. Must call Clone() before using this method.
+	// Meant to be used as the [SetObjectStatus] method for a reconciler.
+	SetReconciliationStatus(reconciler string, status reconciler.Status) Node
+}
+
+type NodeBuilder struct {
+	orig *TableNode
+	new  *TableNode
+}
+
+func (b *NodeBuilder) Build() (Node, error) {
+	if !source.AllowOverwrite(b.orig.Source(), b.new.Source()) {
+		return nil, fmt.Errorf("source %q cannot overwrite %q", b.new.Source(), b.orig.Source())
+	}
+	// TODO add other validations here
+	return b.new, nil
+}
+
+func (b *NodeBuilder) ModifyNode(mod func(n *types.Node)) *NodeBuilder {
+	mod(&b.new.Node)
+	return b
+}
+
+func (b *NodeBuilder) ModifyLocal(mod func(l *LocalNodeAttrs)) *NodeBuilder {
+	if b.new.Local == nil {
+		b.new.Local = &LocalNodeAttrs{}
+	}
+	mod(b.new.Local)
+	return b
+}
+
+// TableNode is the concrete struct stored in the node table.
+// This is exported and JSON serializable for cilium-dbg.
+type TableNode struct {
+	Node     types.Node                   `json:"node"`
+	Local    *LocalNodeAttrs              `json:"local",omitempty`
+	Statuses map[string]reconciler.Status `json:"statuses",omitempty`
+}
+
+var emptyStatusMap = map[string]reconciler.Status{}
+
+func NewTableNode(n types.Node, l *LocalNodeAttrs) Node {
+	return &TableNode{Node: n, Local: l, Statuses: emptyStatusMap}
+}
+
+func (n *TableNode) Clone() Node {
+	return &TableNode{
+		Node:     n.Node,
+		Local:    n.Local,
+		Statuses: n.Statuses,
+	}
+}
+
+func (n *TableNode) GetReconciliationStatus(rc string) reconciler.Status {
+	status, ok := n.Statuses[rc]
+	if ok {
+		return status
+	}
+	// If no existing status, then we assume it's pending. This allows
+	// just setting the map to empty to indicate pending.
+	return reconciler.StatusPending()
+}
+
+func (n *TableNode) SetReconciliationStatus(rc string, status reconciler.Status) Node {
+	n.Statuses = maps.Clone(n.Statuses)
+	n.Statuses[rc] = status
+	return n
+}
+
+func (n *TableNode) Builder() *NodeBuilder {
+	return &NodeBuilder{
+		orig: n,
+		new: &TableNode{
+			Node:     *n.Node.DeepCopy(),
+			Local:    n.Local.Clone(),
+			Statuses: emptyStatusMap,
+		},
+	}
+}
+
+func (n *TableNode) Name() string {
+	return n.Node.Name
+}
+
+func (n *TableNode) Source() source.Source {
+	return n.Node.Source
+}
+
+func (n *TableNode) Identity() types.Identity {
+	return n.Node.Identity()
+}
+
+// Cluster implements Node.
+func (n *TableNode) Cluster() string {
+	return n.Node.Cluster
+}
+
+// GetKeyName implements Node.
+func (n *TableNode) GetKeyName() string {
+	return n.Node.GetKeyName()
+}
+
+// GetNodeIP implements Node.
+func (n *TableNode) GetNodeIP(ipv6 bool) net.IP {
+	return n.Node.GetNodeIP(ipv6)
+}
+
+func (n *TableNode) GetNode() types.Node {
+	return n.Node
+}
+
+func (n *TableNode) GetLocal() *LocalNodeAttrs {
```

Comment:
When I was looking at the way we use `LocalNodeStore` I was kind of noticing these aren't particularly specific to the local node, they're kind of just other fields that the consumers of the node store were using? Not sure if this is needed longer term or if these could just be made into their own getters, or something less "local" oriented.

Reply:
Ah yeah, quite a few of them currently are not specific to the local node (at least UID, ProviderID).  The thing though is that for other nodes we essentially just get `types.Node` and that doesn't carry the UID or ProviderID, so they'd always be zero'd for anything except the local node (well we would get them if in non-kvstore mode, if we change how we process CiliumNode). For that reason I split this up into "remote nodes" that only have `types.Node` as payload, and then the local one which has `types.Node` plus some extra attributes. Not sure what's the right way to slice it.

The design of this definitely still feels half-baked. E.g. should we even use `types.Node` here as it's really geared towards serialization to kvstore, maybe an internal model would be better and we'd convert to `types.Node` when serializing? The `Node` interface and the `LocalNodeAttrs` stuff doesn't feel as lightweight as it should. etc.